examples/fidelity.py

```diff
@@ -0,0 +1,109 @@
+"""Simulates the fidelity estimation.
+
+Direct Fidelity Estimation from Few Pauli Measurements
+https://arxiv.org/abs/1104.4695
+
+Practical characterization of quantum devices without tomography
+https://arxiv.org/abs/1104.3835
+"""
+
+import cirq
+import heapq
+import itertools
+import math
+import numpy
+
+
+def build_circuit():
+  qubits = [cirq.LineQubit(i) for i in range(3)]
+
+  rot0 = cirq.ZPowGate(exponent=0.250)  # T-Gate, non Clifford.
+  rot1 = cirq.XPowGate(exponent=0.123)
+  rot2 = cirq.XPowGate(exponent=0.456)
+
+  circuit = cirq.Circuit()
+  circuit.append(rot0(cirq.LineQubit(0)))
+  circuit.append(rot1(cirq.LineQubit(1)))
+  circuit.append(rot2(cirq.LineQubit(2)))
+
+  return circuit, qubits
+
+
+def build_noisy_circuit(circuit, qubits):
+  noisy_circuit = circuit.copy()
+
+  noisy_gate = cirq.amplitude_damp(0.01)
+  for qubit in qubits:
+    noisy_circuit.append(noisy_gate(qubit))
+
+  return noisy_circuit
+
+
+def simulate_trace(circuit, Pi):
+  simulator = cirq.DensityMatrixSimulator()
+
+  n = len(Pi)
+  d = 2**n
+
+  rot = numpy.asarray([1], numpy.complex64)
+  for op in Pi:
+    if op == 'I':
+      rot = numpy.kron(rot, numpy.asarray([[1, 0], [0, 1]], numpy.complex64))
+    elif op == 'X':
+      rot = numpy.kron(rot, numpy.asarray([[0, 1], [1, 0]], numpy.complex64))
+    elif op == 'Y':
+      rot = numpy.kron(rot, numpy.asarray([[0, -1j], [1j, 0]], numpy.complex64))
+    elif op == 'Z':
+      rot = numpy.kron(rot, numpy.asarray([[1, 0], [0, -1]], numpy.complex64))
+
+  trace = 0
+  for x in range(d):
+    xbin = numpy.binary_repr(x, width=n)
+
+    xvec = numpy.zeros([d], numpy.complex64)
+    xvec[x] = 1
+
+    initial_state = numpy.matmul(rot, xvec)
```

Comment:
```
n = len(pauli_gates)
trace = 0
for x in range(2**n):
    initial_state = cirq.final_wavefunction(
        cirq.DensePauliString(pauli_gates),
        initial_state=x)
    ...
```

Reply:
This is a great function to have. Thanks for the tip. I did try the code as you suggested, but it failed.

I left the print debug statements and I am quoting the _sorted_ log below. There are two weird things:

- Having an cirq.I gate reduces the size of 'rotated_initial_state'
- It seems the output is order-independent, for example IIZ seems to be the same as IZI.

`
TONYBOOM rotated_initial_state=[0.-0.j 0.+0.j 0.-0.j 0.+0.j 0.-0.j 0.+0.j 0.-0.j 0.+1.j] pauli_gates=[cirq.X, cirq.X, cirq.Y]
TONYBOOM rotated_initial_state=[0.-0.j 0.-0.j 0.+0.j 0.+0.j 0.-0.j 0.-0.j 0.+0.j 0.+1.j] pauli_gates=[cirq.X, cirq.Y, cirq.X]
TONYBOOM rotated_initial_state=[0.-0.j 0.-0.j 0.-0.j 0.-0.j 0.+0.j 0.+0.j 0.+0.j 0.+1.j] pauli_gates=[cirq.Y, cirq.X, cirq.X]
TONYBOOM rotated_initial_state=[ 0.+0.j -0.+0.j  0.-0.j  0.+0.j -0.+0.j  0.+0.j  0.-0.j -0.-1.j] pauli_gates=[cirq.Y, cirq.Y, cirq.Y]
TONYBOOM rotated_initial_state=[0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 0.+0.j 1.+0.j] pauli_gates=[cirq.X, cirq.X, cirq.X]
TONYBOOM rotated_initial_state=[-0.+0.j  0.+0.j  0.-0.j  0.+0.j -0.+0.j  0.+0.j  0.-0.j -1.+0.j] pauli_gates=[cirq.X, cirq.Y, cirq.Y]
TONYBOOM rotated_initial_state=[-0.+0.j  0.+0.j -0.+0.j  0.+0.j  0.-0.j  0.+0.j  0.-0.j -1.+0.j] pauli_gates=[cirq.Y, cirq.X, cirq.Y]
TONYBOOM rotated_initial_state=[-0.+0.j -0.+0.j  0.+0.j  0.+0.j  0.-0.j  0.-0.j  0.+0.j -1.+0.j] pauli_gates=[cirq.Y, cirq.Y, cirq.X]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.+0.j -0.+0.j  0.-0.j  0.+0.j  0.+1.j -0.+0.j] pauli_gates=[cirq.X, cirq.Y, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.-0.j  0.+0.j  0.+0.j -0.+0.j  0.+1.j -0.+0.j] pauli_gates=[cirq.Y, cirq.X, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.+0.j -0.+0.j  0.+0.j -0.+0.j  0.+0.j -0.+0.j  1.+0.j -0.+0.j] pauli_gates=[cirq.X, cirq.X, cirq.Z]
TONYBOOM rotated_initial_state=[-0.+0.j  0.-0.j  0.+0.j -0.+0.j  0.-0.j  0.+0.j -1.+0.j -0.+0.j] pauli_gates=[cirq.Y, cirq.Y, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.+0.j -0.+0.j  0.-0.j  0.+1.j  0.+0.j -0.+0.j] pauli_gates=[cirq.X, cirq.Z, cirq.Y]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.-0.j  0.+0.j  0.+0.j  0.+0.j  0.+1.j -0.+0.j -0.+0.j] pauli_gates=[cirq.Y, cirq.Z, cirq.X]
TONYBOOM rotated_initial_state=[ 0.+0.j  0.+0.j -0.+0.j -0.+0.j  0.+0.j  1.+0.j -0.+0.j -0.+0.j] pauli_gates=[cirq.X, cirq.Z, cirq.X]
TONYBOOM rotated_initial_state=[-0.+0.j  0.+0.j  0.-0.j  0.+0.j  0.-0.j -1.+0.j  0.+0.j -0.+0.j] pauli_gates=[cirq.Y, cirq.Z, cirq.Y]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.+0.j -0.+0.j  0.+1.j -0.+0.j -0.+0.j  0.-0.j] pauli_gates=[cirq.Y, cirq.Z, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.+0.j -0.+0.j -0.+0.j  0.-0.j  1.+0.j -0.+0.j -0.+0.j  0.-0.j] pauli_gates=[cirq.X, cirq.Z, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.-0.j  0.+1.j  0.+0.j -0.+0.j  0.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.X, cirq.Y]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.-0.j  0.+0.j  0.+1.j  0.+0.j  0.+0.j -0.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.Y, cirq.X]
TONYBOOM rotated_initial_state=[0.-0.j 0.+0.j 0.-0.j 0.+1.j] pauli_gates=[cirq.I, cirq.X, cirq.Y]
TONYBOOM rotated_initial_state=[0.-0.j 0.-0.j 0.+0.j 0.+1.j] pauli_gates=[cirq.I, cirq.Y, cirq.X]
TONYBOOM rotated_initial_state=[0.-0.j 0.+0.j 0.-0.j 0.+1.j] pauli_gates=[cirq.X, cirq.I, cirq.Y]
TONYBOOM rotated_initial_state=[0.-0.j 0.+0.j 0.-0.j 0.+1.j] pauli_gates=[cirq.X, cirq.Y, cirq.I]
TONYBOOM rotated_initial_state=[0.-0.j 0.-0.j 0.+0.j 0.+1.j] pauli_gates=[cirq.Y, cirq.I, cirq.X]
TONYBOOM rotated_initial_state=[0.-0.j 0.-0.j 0.+0.j 0.+1.j] pauli_gates=[cirq.Y, cirq.X, cirq.I]
TONYBOOM rotated_initial_state=[ 0.+0.j  0.+0.j  0.+0.j  1.+0.j -0.+0.j -0.+0.j -0.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.X, cirq.X]
TONYBOOM rotated_initial_state=[-0.+0.j  0.+0.j  0.-0.j -1.+0.j  0.-0.j  0.+0.j  0.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.Y, cirq.Y]
TONYBOOM rotated_initial_state=[0.+0.j 0.+0.j 0.+0.j 1.+0.j] pauli_gates=[cirq.I, cirq.X, cirq.X]
TONYBOOM rotated_initial_state=[-0.+0.j  0.+0.j  0.-0.j -1.+0.j] pauli_gates=[cirq.I, cirq.Y, cirq.Y]
TONYBOOM rotated_initial_state=[0.+0.j 0.+0.j 0.+0.j 1.+0.j] pauli_gates=[cirq.X, cirq.I, cirq.X]
TONYBOOM rotated_initial_state=[0.+0.j 0.+0.j 0.+0.j 1.+0.j] pauli_gates=[cirq.X, cirq.X, cirq.I]
TONYBOOM rotated_initial_state=[-0.+0.j  0.+0.j  0.-0.j -1.+0.j] pauli_gates=[cirq.Y, cirq.I, cirq.Y]
TONYBOOM rotated_initial_state=[-0.+0.j  0.+0.j  0.-0.j -1.+0.j] pauli_gates=[cirq.Y, cirq.Y, cirq.I]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.+1.j -0.+0.j  0.+0.j -0.+0.j -0.+0.j  0.-0.j] pauli_gates=[cirq.Z, cirq.Y, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.+1.j -0.+0.j] pauli_gates=[cirq.I, cirq.Y, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.+1.j -0.+0.j] pauli_gates=[cirq.Y, cirq.I, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+0.j  0.+1.j -0.+0.j] pauli_gates=[cirq.Y, cirq.Z, cirq.I]
TONYBOOM rotated_initial_state=[ 0.+0.j -0.+0.j  1.+0.j -0.+0.j -0.+0.j  0.-0.j -0.+0.j  0.-0.j] pauli_gates=[cirq.Z, cirq.X, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.+0.j -0.+0.j  1.+0.j -0.+0.j] pauli_gates=[cirq.I, cirq.X, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.+0.j -0.+0.j  1.+0.j -0.+0.j] pauli_gates=[cirq.X, cirq.I, cirq.Z]
TONYBOOM rotated_initial_state=[ 0.+0.j -0.+0.j  1.+0.j -0.+0.j] pauli_gates=[cirq.X, cirq.Z, cirq.I]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+1.j  0.+0.j -0.+0.j  0.+0.j -0.+0.j -0.+0.j  0.+0.j] pauli_gates=[cirq.Z, cirq.Z, cirq.Y]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+1.j  0.+0.j -0.+0.j] pauli_gates=[cirq.I, cirq.Z, cirq.Y]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+1.j  0.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.I, cirq.Y]
TONYBOOM rotated_initial_state=[ 0.-0.j  0.+1.j  0.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.Y, cirq.I]
TONYBOOM rotated_initial_state=[0.-0.j 0.+1.j] pauli_gates=[cirq.I, cirq.I, cirq.Y]
TONYBOOM rotated_initial_state=[0.-0.j 0.+1.j] pauli_gates=[cirq.I, cirq.Y, cirq.I]
TONYBOOM rotated_initial_state=[0.-0.j 0.+1.j] pauli_gates=[cirq.Y, cirq.I, cirq.I]
TONYBOOM rotated_initial_state=[ 0.+0.j  1.+0.j -0.+0.j -0.+0.j -0.+0.j -0.+0.j  0.-0.j  0.-0.j] pauli_gates=[cirq.Z, cirq.Z, cirq.X]
TONYBOOM rotated_initial_state=[ 0.+0.j  1.+0.j -0.+0.j -0.+0.j] pauli_gates=[cirq.I, cirq.Z, cirq.X]
TONYBOOM rotated_initial_state=[ 0.+0.j  1.+0.j -0.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.I, cirq.X]
TONYBOOM rotated_initial_state=[ 0.+0.j  1.+0.j -0.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.X, cirq.I]
TONYBOOM rotated_initial_state=[0.+0.j 1.+0.j] pauli_gates=[cirq.I, cirq.I, cirq.X]
TONYBOOM rotated_initial_state=[0.+0.j 1.+0.j] pauli_gates=[cirq.I, cirq.X, cirq.I]
TONYBOOM rotated_initial_state=[0.+0.j 1.+0.j] pauli_gates=[cirq.X, cirq.I, cirq.I]
TONYBOOM rotated_initial_state=[ 1.+0.j -0.+0.j -0.+0.j  0.-0.j -0.+0.j  0.-0.j  0.-0.j  0.+0.j] pauli_gates=[cirq.Z, cirq.Z, cirq.Z]
TONYBOOM rotated_initial_state=[ 1.+0.j -0.+0.j -0.+0.j  0.-0.j] pauli_gates=[cirq.I, cirq.Z, cirq.Z]
TONYBOOM rotated_initial_state=[ 1.+0.j -0.+0.j -0.+0.j  0.-0.j] pauli_gates=[cirq.Z, cirq.I, cirq.Z]
TONYBOOM rotated_initial_state=[ 1.+0.j -0.+0.j -0.+0.j  0.-0.j] pauli_gates=[cirq.Z, cirq.Z, cirq.I]
TONYBOOM rotated_initial_state=[ 1.+0.j -0.+0.j] pauli_gates=[cirq.I, cirq.I, cirq.Z]
TONYBOOM rotated_initial_state=[ 1.+0.j -0.+0.j] pauli_gates=[cirq.I, cirq.Z, cirq.I]
TONYBOOM rotated_initial_state=[ 1.+0.j -0.+0.j] pauli_gates=[cirq.Z, cirq.I, cirq.I]
TONYBOOM rotated_initial_state=[1.+0.j] pauli_gates=[cirq.I, cirq.I, cirq.I]
`

Reply:
I did look at the code for parameters, or a bu maybe, but I can't find anything obvious.

Leaving as not resolved.